File: src/data/types/buildCitadelResult.ts
import type { Wallet } from "./wallet.ts";
import type { PlanetOwnershipEntry } from "./planetOwnershipEntry.ts";

// buildCitadel() never touches Inventory directly -- same boundary craft()
// already holds (it takes pre-resolved ResourceInstance[], never the
// Inventory container itself). The caller checks/consumes materials from
// its own inventory state after a successful result, exactly like
// MarketScene calls removeBatchAt() after sellToMarket() succeeds.
export interface BuildCitadelSucceeded {
  success: true;
  updatedWallet: Wallet;
  updatedOwnershipEntry: PlanetOwnershipEntry;
  materialResourceId: string | null;
  materialQuantityConsumed: number;
}

export interface BuildCitadelRejected {
  success: false;
  reason: string;
}

export type BuildCitadelResult = BuildCitadelSucceeded | BuildCitadelRejected;

File: unity/ProfitableCore/Schema/BuildCitadelResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/buildCitadelResult.ts's BuildCitadelResult =
// BuildCitadelSucceeded | BuildCitadelRejected discriminated union.
// BuildCitadel never touches Inventory directly -- same boundary
// Crafter.Craft already holds; the caller checks/consumes materials from
// its own inventory state after a successful result.
public abstract class BuildCitadelResult
{
    public abstract bool Success { get; }
}

public sealed class BuildCitadelSucceeded : BuildCitadelResult
{
    public override bool Success => true;
    public Wallet UpdatedWallet { get; init; } = new();
    public PlanetOwnershipEntry UpdatedOwnershipEntry { get; init; } = new();
    public string? MaterialResourceId { get; init; }
    public int MaterialQuantityConsumed { get; init; }
}

public sealed class BuildCitadelRejected : BuildCitadelResult
{
    public override bool Success => false;
    public string Reason { get; init; } = string.Empty;
}
